File: python3/25.py
#! /usr/bin/env python3

card_pubkey = 9033205
door_pubkey = 9281649

card_loopsize = 0

def transform(subjnum, loopsize):
    val = 1
    for i in range(loopsize):
        val *= subjnum
        val %= 20201227
    print(loopsize)
    return val

# find card_loopsize: transform 7 until card_pubkey achieved
val = 1
subjnum = 7
for ls in range(1,20000000):
    val *= subjnum
    val %= 20201227

    if val == card_pubkey:
        card_loopsize = ls
        print('cls =', card_loopsize)
        break
else:
    print('NF')

# part 1 - 9714832
encryption_key = transform(door_pubkey, card_loopsize)
print('p1', encryption_key)
print(transform(1,3))

File: go/25.go
package main

import (
	"fmt"
)

func transform(subjnum uint64, loopsize uint64) uint64 {
	val := uint64(1)
	for i := uint64(0); i < loopsize; i++ {
		val *= subjnum
		val %= 20201227
	}
	return val
}

func main() {
	card_pubkey := uint64(9033205)
	door_pubkey := uint64(9281649)
	card_loopsize := uint64(0)

	val := uint64(1)
	subjnum := uint64(7)
	for cls := uint64(1); cls < 20000000; cls++ {
		val *= subjnum
		val %= 20201227

		if val == card_pubkey {
			card_loopsize = cls
			fmt.Printf("cls = %d\n", card_loopsize)
		}
	}

	// part 1 - 9714832
	encryption_key := transform(door_pubkey, card_loopsize)
	fmt.Println("p1", encryption_key)
}
